### ISSUE TITLE
Some sort of corruption is occurring with permessage-deflate

### ISSUE THREAD

**Issue by glasser** [faye/permessage-deflate-node#4]
(I apologize for a not particularly helpful bug report, but I think some report is better than nothing.)

I'm trying to enable permessage-deflate by default in Meteor 1.0.4, but I'm seeing definite signs that some level of message corruption is occurring in our test suite.  It does not happen every time I run our test suite, but under some circumstances that I have not yet tracked down, messages are getting blatantly corrupted.  The client in this case is Chrome.

I am running `faye-websocket` 0.9.3, `websocket-driver` 0.5.3, `websocket-extensions` 0.1.1, and `permessage-deflate` 0.1.2.

**Comment by glasser** [comment]
I have yet to make a small reproduction, but this does appear to be fixed with `noContextTakeover` on the server. So it seems like it could be some sort of concurrency issue.

**Comment by glasser** [comment]
And note that it isn't like it is immediately broken when it breaks. It manages to send hundreds of messages... and then something goes bad and they end up mangled.

![permessage-deflate](https://cloud.githubusercontent.com/assets/16724/6565536/ee4e9306-c66d-11e4-8c06-9153cca3113c.png)

**Comment by glasser** [comment]
And oddly, the first broken message is broken by having the last 7 bytes (`}]}}}"]`) moved from the end to the beginning.

**Comment by jcoglan** [comment]
I'm not sure what's going on here, given this passes all the Autobahn tests that transfer a lot of data. The fact that the issue does not occur with `server_no_context_takeover` means that there must be a problem with how the reused server-side deflate session is being accessed. When using context takeover, the lock/queue mechanism in this module is supposed to make sure that only one message at a time is using the deflate session.

Given the output, it seems like the compressed data is being emitted to the socket in the wrong order, either within a message/frame or across them. Some possible causes:
- The lock/queue mechanism in this module, for blocking concurrent use of reused sessions, has a bug
- The compressed chunks emitted by this module are reordered within a message
- Messages emitted by this module are written to the wire in the wrong order
- There is an issue with the [extension pipeline](https://github.com/faye/websocket-extensions-node/tree/master/lib/pipeline), for example maybe the ring buffer implementation there is not preserving order correctly

What I suggest you do is log the server-side activity of your sockets, by loading this patch. You might need to modify the module paths for your setup.

``` js
var crypto     = require('crypto'),
    fs         = require('fs'),
    Extensions = require('websocket-driver/node_modules/websocket-extensions'),
    Session    = require('permessage-deflate/lib/session');


Extensions.prototype._getId = function() {
  return this._id = this._id || crypto.randomBytes(8).toString('hex');
};


var ExtensionsProcessIncomingMessage = Extensions.prototype.processIncomingMessage;

Extensions.prototype.processIncomingMessage = function(message, callback, context) {
  this._inLog = this._inLog || fs.createWriteStream(this._getId() + '.in.log');
  this._inSeq = this._inSeq || [0, 0];

  message.id = ++this._inSeq[0];
  var self   = this;

  ExtensionsProcessIncomingMessage.call(this, message, function(error, result) {
    if (result) {
      result._log[0].push(++self._inSeq[1]);
      self._inLog.write(JSON.stringify(result._log) + '\n');
    }
    callback.call(context, error, result);
  });
};


var ExtensionsProcessOutgoingMessage = Extensions.prototype.processOutgoingMessage;

Extensions.prototype.processOutgoingMessage = function(message, callback, context) {
  this._outLog = this._outLog || fs.createWriteStream(this._getId() + '.out.log');
  this._outSeq = this._outSeq || [0, 0];

  message.id = ++this._outSeq[0];
  var self   = this;

  ExtensionsProcessOutgoingMessage.call(this, message, function(error, result) {
    if (result) {
      result._log[0].push(++self._outSeq[1]);
      self._outLog.write(JSON.stringify(result._log) + '\n');
    }
    callback.call(context, error, result);
  });
};


var SessionProcessIncomingMessage = Session.prototype.processIncomingMessage;

Session.prototype.processIncomingMessage = function(message, callback) {
  this._inSeq = this._inSeq || 0;

  var id    = message.id,
      data  = message.data.toString('base64'),
      rsv1  = message.rsv1,
      count = ++ this._inSeq;

  SessionProcessIncomingMessage.call(this, message, function(error, result) {
    if (result) {
      var output = result.data.toString('base64');
      result._log = [[id], count, [rsv1], data, output];
    }
    callback(error, result);
  });
};


var SessionProcessOutgoingMessage = Session.prototype.processOutgoingMessage;

Session.prototype.processOutgoingMessage = function(message, callback) {
  this._outSeq = this._outSeq || 0;

  var id    = message.id,
      data  = message.data.toString('base64'),
      count = ++this._outSeq;

  SessionProcessOutgoingMessage.call(this, message, function(error, result) {
    if (result) {
      var output = result.data.toString('base64');
      result._log = [[id], count, [result.rsv1], data, output];
    }
    callback(error, result);
  });
};


var close = Session.prototype.close;

Session.prototype.close = function() {
  if (this._inLog) this._inLog.close();
  if (this._outLog) this._outLog.close();
  close.call(this);
};
```

For every socket, this produces two logfiles, one for incoming and one for outgoing data. Each log line records these data:
- A pair of sequence numbers representing the order in which the message entered and left the `Extensions` pipeline; these should always match
- A sequence number representing the order in which `permessage-deflate` received the message; this should match the previous sequence numbers
- The set of `rsv` flags
- The received message payload data as base64
- The produced message payload data (after inflation/deflation) as base64

Recording the data payloads means you could replay them against a deflate stream yourself and see if something different happens.

Run your examples using this patch and let me know what happens.

**Comment by glasser** [comment]
Yeah, I'm sorry for the worthless bug report.  Caught this as part of pre-release QA, and decided then that the best use of time was to just revert use of the new code rather than investigate.  But I still really want to use it, and will give your patch (which is a better version of something I was imagining doing) a shot at some point hopefully soon.

**Comment by jcoglan** [comment]
I just realised that `close()` patch should be on `Extensions` rather than `Session`; I suppose you can figure this out, I just wanted to give you something to start from.

It occurs to me that [this pattern](https://github.com/faye/permessage-deflate-node/blob/037afc350cb62aac1583b8395423c7a5d5838d66/lib/session.js#L85-L89) might be a problem:

``` js
    self._lockOut = false;
    var next = self._queueOut.shift();
    if (next) self.processOutgoingMessage.apply(self, next);

    callback(error, message);
```

i.e. calling back into the code that handles the current message _after_ popping the next message off the queue. If something happens synchronously where I wasn't expect it, this might lead to out-of-order processing. But I'd need you to repro and test this idea with your stack.

**Comment by jcoglan** [comment]
A similar report from Alistair Braidwood:

> I have a simple server that does this:
> 
> ``` js
> bayeux.on('subscribe', function(clientId, channel) {
>   var ccypair = channel.replace(/(^\/FX\/PRICING\/|\/.*$)/g, '');
> 
>   PriceGen.addListener(ccypair, function(update) {
>     bayeux.getClient().publish(channel, update);
>   });
> });
> ```
> 
> and a client that makes 8 subscriptions on different channels.
> 
> The server publishes at a rate of >100/sec with varying message sizes.
> 
> After a while, I get 2 types of errors thrown in Chrome, Firefox and Safari on OS X 10.10.  One type has an empty `event.data` in `socket.onmessage`, the other has garbage in the JSON data (something like `}]][{"channel":...` which looks like a buffer is wrapping around and struggling with previous data.

**Comment by glasser** [comment]
OK, I have a dump generated in the wild.  Note that everything below is with
Node 0.10.36.

I made a change to your patch, changing processOutgoingMessage to

```
Session.prototype.processOutgoingMessage = function(message, callback) {
  if (this._lockOut) {
    return SessionProcessOutgoingMessage.call(this, message, callback);
  }

  this._outSeq = this._outSeq || 0;

  var id    = message.id,
      dataOrig = message.data.toString('utf8'),
      data  = message.data.toString('base64'),
      count = ++this._outSeq;

  SessionProcessOutgoingMessage.call(this, message, function(error, result) {
    if (result) {
      var output = result.data.toString('base64');
      result._log = [[id], count, [result.rsv1], data, dataOrig, output];
    }
    callback(error, result);
  });
};
```

The changes are including dataOrig (for ease of debugging; all my messages are
valid UTF8), and the conditional at the top. Without the conditional, the patch
(including _outSeq increment) would execute both on the "real" execution of
processOutgoingMessage and on the one that just pushes it onto the queue.
(Incoming has the same issue but I was only focusing on outgoing.)

So, here is a (very long) dump that's broken.

I ran the following script to find issues:

``` js
var fs = require('fs');
var assert = require('assert');
var deflate = require('./packages/ddp/.npm/package/node_modules/permessage-deflate').configure({});
var serverSession = deflate.createServerSession([{}]);
var clientSession = deflate.createClientSession();
clientSession.activate({});
var records = fs.readFileSync('bad.out.log', 'utf8').split('\n');
records.pop();

function expandProcess (i) {
  if (i >= records.length)
    return;
  var record = JSON.parse(records[i]);

  var message = { data: new Buffer(record[5], 'base64'), rsv1: true };
  clientSession.processIncomingMessage(message, function (err, processed) {
    if (err) throw err;
    var expected = new Buffer(record[3], 'base64');
    var actual = processed.data;
    try {
      assert.deepEqual(actual, expected);
    } catch (e) {
      console.log("expand doesn't match", records[i]);
      return;
    }
    expandProcess(i + 1);
  });
};

function compressProcess (i) {
  if (i >= records.length)
    return;
  var record = JSON.parse(records[i]);

  var message = { data: new Buffer(record[3], 'base64'), DEBUG: record[0][0] === 3337 || record[0][0] === 3336 };
  clientSession.processOutgoingMessage(message, function (err, processed) {
    if (err) throw err;
    var expected = new Buffer(record[5], 'base64');
    var actual = processed.data;
    try {
      assert.deepEqual(actual, expected);
    } catch (e) {
      console.log("compress not replicable", records[i]);
      return;
    }

    compressProcess(i + 1);
  });
};

expandProcess(0);
compressProcess(0);
```

Output:

```
compress not replicable [[3764,3764],3764,[true],"YVsie1wibXNnXCI6XCJjaGFuZ2VkXCIsXCJjb2xsZWN0aW9uXCI6XCJ0aW55dGVzdF9yZXN1bHRzX2NvbGxlY3Rpb25cIixcImlkXCI6XCJqTWZ6anZTZWI5RTRqbzlMelwiLFwiZmllbGRzXCI6e1wiNWNKUkp3QmR1Y252UHZvdWlcIjp7XCJncm91cFBhdGhcIjpbXCJ0aW55dGVzdFwiLFwiaHR0cGNhbGxcIl0sXCJ0ZXN0XCI6XCJ0aW1lb3V0XCIsXCJldmVudHNcIjpbe1wic2VxdWVuY2VcIjozMjQ5LFwidHlwZVwiOlwib2tcIn1dfX19Il0=","a[\"{\\\"msg\\\":\\\"changed\\\",\\\"collection\\\":\\\"tinytest_results_collection\\\",\\\"id\\\":\\\"jMfzjvSeb9E4jo9Lz\\\",\\\"fields\\\":{\\\"5cJRJwBducnvPvoui\\\":{\\\"groupPath\\\":[\\\"tinytest\\\",\\\"httpcall\\\"],\\\"test\\\":\\\"timeout\\\",\\\"events\\\":[{\\\"sequence\\\":3249,\\\"type\\\":\\\"ok\\\"}]}}}\"]","GoAiL9kryKvcKaU0Oa8soCy/lPxqDuSZ/NISQmGBq1UHAA=="]
expand doesn't match [[3337,3337],3337,[true],"YVsie1wibXNnXCI6XCJjaGFuZ2VkXCIsXCJjb2xsZWN0aW9uXCI6XCJ0aW55dGVzdF9yZXN1bHRzX2NvbGxlY3Rpb25cIixcImlkXCI6XCJqTWZ6anZTZWI5RTRqbzlMelwiLFwiZmllbGRzXCI6e1wiMkxTcXVDeFM3V1dwRHRiSnVcIjp7XCJncm91cFBhdGhcIjpbXCJ0aW55dGVzdFwiLFwiaHRtbC10b29sc1wiXSxcInRlc3RcIjpcInBhcnNlciBnZXRDb250ZW50XCIsXCJldmVudHNcIjpbe1wic2VxdWVuY2VcIjoyODIyLFwidHlwZVwiOlwib2tcIn1dfX19Il0=","a[\"{\\\"msg\\\":\\\"changed\\\",\\\"collection\\\":\\\"tinytest_results_collection\\\",\\\"id\\\":\\\"jMfzjvSeb9E4jo9Lz\\\",\\\"fields\\\":{\\\"2LSquCxS7WWpDtbJu\\\":{\\\"groupPath\\\":[\\\"tinytest\\\",\\\"html-tools\\\"],\\\"test\\\":\\\"parser getContent\\\",\\\"events\\\":[{\\\"sequence\\\":2822,\\\"type\\\":\\\"ok\\\"}]}}}\"]","AA=="]
```

This means that element 3337's compression was wrong, both during the original
run of the program, and during the check script.  However, the check script
managed to have identical output to the original program up until 3764: ie, the
check script is reproducing the erroneous compression.  The compression is
clearly wrong, since the output "AA==" is `<Buffer 00>`, which sure seems
unlikely for a non-trivial input.

I dug into it and it looks like a bug (or at least poor design) in Node's zlib
wrapper.  As far as I can tell, every low-level call to `deflate` with
Z_SYNC_FLUSH leaves the 0000ffff on the end.  So when `deflate` runs out of room
in its output buffer, it will still return an output buffer to Node that ends in
0000ffff, even though the Node wrapper still needs to call `deflate` again to
continue processing the input!  (This is happening in
`Zlib.prototype._processChunk`; each call to `this._handle.write`
(`this._binding.write` in Node 0.10) corresponds to a `deflate` call.)

So sometimes (specifically, on element 3336 of this file), even after
permessage-deflate decides that a block is over, zlib.js still may have another
bit of data coming.  In this particular case it seems to be data that
decompresses to nothing.  (Possibly in other cases it is data that decompresses
to something.) And when the 'data' handler is added for the _next_ message, it
gets the `<00 00 00 ff ff>` that is queued up and decides that this is the next
message.

Not really clear how to fix this.  It's arguably a Node bug?  It seems to mean
that the output of a DeflateRaw stream is actually dependent on the chunkSize
chosen for compression (a Node thing, not a zlib thing).  Arguably if you're
using Z_SYNC_FLUSH (or Z_FULL_FLUSH) zlib.js should be stripping internal
0000ffff's from the stream...

Here's a simple way to demonstrate the Node issue:

``` js
var zlib = require('zlib');
var deflate = zlib.createDeflateRaw({
  flush: zlib.Z_SYNC_FLUSH,
  chunkSize: +process.env.CS || zlib.Z_DEFAULT_CHUNK
});

deflate.pipe(process.stdout);
var text = '';
for (var i = 0; i < 1000; ++i) {
  text += i + '\n';
}
deflate.write(text);
```

```
$ node zlib-test.js | wc -c
    1833
$ CS=100 node zlib-test.js | wc -c
    1833
$ CS=1830 node zlib-test.js | wc -c
    1838
$ CS=1833 node zlib-test.js | wc -c
    1838
$ CS=1835 node zlib-test.js | wc -c
    1833
```

ie, the output of this deflate stream depends on the chunk size!

**Comment by glasser** [comment]
Though I wonder, is it possible that we are not implementing

> 1.  If the resulting data does not end with an empty DEFLATE block
>      with no compression (the "BTYPE" bits are set to 00), append an
>      empty DEFLATE block with no compression to the tail end.

from the permessage-deflate RFC properly?  ie, maybe this particular case can be detected because the last block is not empty?

**Comment by jcoglan** [comment]
Could you implement that RFC point you mention and see if it makes a difference?

Do you have the full log of your session? I didn't see a link in your message.

**Comment by glasser** [comment]
Er, whoops. https://gist.githubusercontent.com/glasser/b58328e2b24a4fdfba7b/raw/bad.out.log

So yeah, in short, what I think is going on here is a weird property of Z_SYNC_FLUSH (at the C level) where it works this way.  If the empty non-compressed block (00 00 00 ff ff) will go off the edge of `deflate`'s output buffer, or align with it exactly, you need to call `deflate` again, because you're supposed to always call it again if `deflate` ends with 0 output space available.  But it also tries to always include a full non-compressed block in the last call to `deflate`.  So this implies that sometimes you'll end up with two non-compressed blocks at the end.

If you were just calling `deflate` manually yourself this wouldn't be a big deal --- you'd know when you were done because you'd be running that loop (keep `deflate`ing until you get non-zero available output space) yourself.  And I _think_ it's perfectly legit for you to include extra empty blocks in your permessage-deflate messages (as long as each message ends with a "truncated" (no 00 00 ff ff) empty block).

But in the streaming setup that Node has, I'm not sure if there's a great way to tell that all the stuff you've put onto the DeflateRaw stream has been processed --- ie, that a single call to `_transform` has reached `cb`.

I suppose you could do something nuts like subclass `DeflateRaw`, override `push` and some other stuff, and manually invoke `_transform` with your own callback, instead of actually using the stream API.

**Comment by glasser** [comment]
BTW, here's another way of thinking about this:  The empty uncompressed block is _not_ intended by `deflate` as a sentinel.  It wouldn't make sense for it to be, for two reasons:
- It's a perfectly legitimate part of a deflate stream, just a kind of silly one if you assume the goal of compression is to not include a bunch of bytes that don't affect the output.
- The C-level `deflate` API doesn't need a sentinel to tell you when it's done; you're just supposed to keep running it until it finishes a call without consuming the entire output buffer.

The actual purpose of this block is to align the bit-based deflate stream on a byte boundary.

Unfortunately, since Node hides the call to `deflate` inside its stream, you don't get to use `deflate`'s sentinel.

You could just call `deflate.flush` and use its callback to decide when you are done (looks like you call it, though ignore its callback, in Node 0.8 already).  This would have the minor issue that it would often include a few extra bytes of an empty encoding; worse, though, in Node 0.10 it always uses Z_FULL_FLUSH which is like Z_SYNC_FLUSH but also resets the compression state (ie, makes your compression worse... you might as well just not reuse context).  In Node 0.12 you can specify a specific flush flag (eg Z_SYNC_FLUSH) though.

(One probably-bad idea would be to just ignore `require('zlib')` and go directly to `process.binding('zlib')`, which is the underlying handle to the C++ code.  It actually has a pretty minimal interface and wouldn't be hard to write, but it's probably not something you want to rely on.)

So I think short of getting a change into Node (or using undocumented capabilities), the best thing to do would be to always call `deflate.flush` and use its callback instead of the bytes to determine when you are done.  Detect if you're on 0.12 or newer and in that case pass `Z_SYNC_FLUSH` to `flush`.  And if you want to be truly optimized, instead of replacing a single trailing instance of `\x00\0x00\x00\xff\xff` with '\x00', replace multiple trailing instances of that string with one `\x00` (since I think the flush call is likely to add another).

**Comment by glasser** [comment]
Actually, I think the right way to avoid the extraneous blocks is just to not initialize the deflate with a `flush` option (so that the `write`s don't generate ending blocks, only the `flush`).

**Comment by jcoglan** [comment]
I ran your script against the root of this repo and it produces the result you describe. I then wrote my own script to check I understood things; this script takes each input from the logs, compresses it, compares it to the output from the logs, and then decompresses it:

``` js
var fs      = require('fs'),
    deflate = require('./');

var records = fs.readFileSync('bad.out.log', 'utf8')
              .replace(/\s*$/g, '')
              .split(/\n/)
              .map(JSON.parse);

var client   = deflate.createClientSession(),
    offer    = client.generateOffer(),
    server   = deflate.createServerSession([offer]),
    response = server.generateResponse();

console.log(offer, response);

client.activate(response);

function check(index) {
  var record = records[index];
  if (!record) return;

  var input   = record[3],
      output  = record[5],
      message = {data: new Buffer(input, 'base64')};

  server.processOutgoingMessage(message, function(error, message) {
    if (message.data.toString('base64') !== output) {
      console.error('Deflate failed on', record);
      console.error(message, message.data.toString('base64'));
      process.exit(1);
    }

    client.processIncomingMessage(message, function(error, message) {
      if (message.data.toString('base64') !== input) {
        console.error('Inflate failed on', record);
        process.exit(1);
      }
      check(index + 1);
    });
  });
}

check(0);
```

This fails on record 3337:

```
{ client_max_window_bits: true } {}
Deflate failed on [ [ 3337, 3337 ],
  3337,
  [ true ],
  'YVsie1wibXNnXCI6XCJjaGFuZ2VkXCIsXCJjb2xsZWN0aW9uXCI6XCJ0aW55dGVzdF9yZXN1bHRzX2NvbGxlY3Rpb25cIixcImlkXCI6XCJqTWZ6anZTZWI5RTRqbzlMelwiLFwiZmllbGRzXCI6e1wiMkxTcXVDeFM3V1dwRHRiSnVcIjp7XCJncm91cFBhdGhcIjpbXCJ0aW55dGVzdFwiLFwiaHRtbC10b29sc1wiXSxcInRlc3RcIjpcInBhcnNlciBnZXRDb250ZW50XCIsXCJldmVudHNcIjpbe1wic2VxdWVuY2VcIjoyODIyLFwidHlwZVwiOlwib2tcIn1dfX19Il0=',
  'a["{\\"msg\\":\\"changed\\",\\"collection\\":\\"tinytest_results_collection\\",\\"id\\":\\"jMfzjvSeb9E4jo9Lz\\",\\"fields\\":{\\"2LSquCxS7WWpDtbJu\\":{\\"groupPath\\":[\\"tinytest\\",\\"html-tools\\"],\\"test\\":\\"parser getContent\\",\\"events\\":[{\\"sequence\\":2822,\\"type\\":\\"ok\\"}]}}}"]',
  'AA==' ]
{ data: <Buffer 1a 80 21 59 9f e0 c2 52 e7 8a 60 f3 f0 f0 02 97 92 24 2f ba 06 03 ae 56 24 00>,
  rsv1: true } 'GoAhWZ/gwlLnimDz8PACl5IkL7oGA65WJAA='
```

I wasn't sure why this was happening, so I tried setting up my server and client session as you had, but got the same result. (Your script also uses the client session to compress and decompress, whereas mine uses the server to compress, but this makes no difference.)

If you remove the check on compressed data, this script round-trips everything correctly.

I replaced the whole `client.processIncomingMessage()` block with this:

``` js
check(index + 1);
```

Now it fails on the same message you saw:

```
{ client_max_window_bits: true } {}
Deflate failed on [ [ 3764, 3764 ],
  3764,
  [ true ],
  'YVsie1wibXNnXCI6XCJjaGFuZ2VkXCIsXCJjb2xsZWN0aW9uXCI6XCJ0aW55dGVzdF9yZXN1bHRzX2NvbGxlY3Rpb25cIixcImlkXCI6XCJqTWZ6anZTZWI5RTRqbzlMelwiLFwiZmllbGRzXCI6e1wiNWNKUkp3QmR1Y252UHZvdWlcIjp7XCJncm91cFBhdGhcIjpbXCJ0aW55dGVzdFwiLFwiaHR0cGNhbGxcIl0sXCJ0ZXN0XCI6XCJ0aW1lb3V0XCIsXCJldmVudHNcIjpbe1wic2VxdWVuY2VcIjozMjQ5LFwidHlwZVwiOlwib2tcIn1dfX19Il0=',
  'a["{\\"msg\\":\\"changed\\",\\"collection\\":\\"tinytest_results_collection\\",\\"id\\":\\"jMfzjvSeb9E4jo9Lz\\",\\"fields\\":{\\"5cJRJwBducnvPvoui\\":{\\"groupPath\\":[\\"tinytest\\",\\"httpcall\\"],\\"test\\":\\"timeout\\",\\"events\\":[{\\"sequence\\":3249,\\"type\\":\\"ok\\"}]}}}"]',
  'GoAiL9kryKvcKaU0Oa8soCy/lPxqDuSZ/NISQmGBq1UHAA==' ]
{ data: <Buffer 1a 80 d4 1e 5e 6e 51 6a 1e 95 1b 55 19 59 66 e2 e3 6d 6e 42 63 9f a3 b4 e1 b0 cd 3d c0 a6 1e 00 00>,
  rsv1: true } 'GoDUHl5uUWoelRtVGVlm4uNtbkJjn6O04bDNPcCmHgAA'
```

If I replace that line with:

``` js
process.nextTick(function() { check(index + 1) });
```

then it still fails on `3764`. However if I do

``` js
setTimeout(function() { check(index + 1) }, 1);
```

then it fails on `3337`. So the `client.processIncomingMessage()` call was affecting the result just by taking time to execute, and delaying the next compression call. I'm not sure if this affects your view of the problem, although it seems like you know a lot more about DEFLATE (the format and the implementation) than I do.

My main problem with writing the code for this was getting code that worked across Node versions and produces optimal output, without a lot of extra padding on the ends of messages. If you know how to fix that within these constraints (use [wstest](https://github.com/faye/wstest) to check) I will happily accept a PR.

I've looked at the `ws` codebase and they use `flush: zlib.Z_SYNC_FLUSH` _and_ call `deflate.flush()`, which according to your notes might be safe but produce bloated frames.

**Comment by jcoglan** [comment]
By the way, I've no idea why delaying the next call makes the checking get stuck on `3337` rather than getting through it successfully. We know all records before that have identical output to the logs, so it must be that compressing `3337` _earlier_ means it gets compressed properly.

Maybe `data` events get lost in the gap while we're not listening to the `deflate` stream.

**Comment by jcoglan** [comment]
In which case, maybe we should never stop listening to `data`, but capture buffers emitted 'between' messages and roll them into the output for the next message. My concern with playing with `flush()` is that I found it very hard to come up with `flush()` usage that worked across Node versions and produced minimal output.

**Comment by jcoglan** [comment]
The more I think about it, the more this seems right: because we don't call `flush()` with a callback, we assume that resulting `data` events are emitted synchronously with that call, and they might not be. Hence, they might be emitted in a 'gap' while we're not listening for them. I'll see if there's a way to fix `flush()` usage without bloating the output.

**Comment by glasser** [comment]
I agree that this is the cause of the race condition, and that using a flush callback should fix it. Kind of funny that it's a race condition where MISSING events is what avoids errors rather than causes them.

The ws code looks correct to me but not optimized since I don't think you need to set the flush flag if you're also manually flushing (and it does not take advantage of 0.12's ability to make flush use Z_SYNC_FLUSH instead of Z_FULL_FLUSH).

**Comment by jcoglan** [comment]
I wrote this script to fully compress all input messages and then inflate them:

``` js
var fs      = require('fs'),
    deflate = require('./');

var records = fs.readFileSync('bad.out.log', 'utf8')
              .replace(/\s*$/g, '')
              .split(/\n/)
              .map(JSON.parse);

var client     = deflate.createClientSession(),
    offer      = client.generateOffer(),
    server     = deflate.createServerSession([offer]),
    response   = server.generateResponse(),
    compressed = [];

client.activate(response);

function compress(index) {
  var record = records[index];
  if (!record) return decompress(0);

  var message = {data: new Buffer(record[3], 'base64')};

  server.processOutgoingMessage(message, function(error, message) {
    compressed[index] = message;
    compress(index + 1);
  });
}

function decompress(index) {
  var record = records[index];
  if (!record) return;

  var payload = record[3],
      message = compressed[index];

  client.processIncomingMessage(message, function(error, message) {
    var output = message.data.toString('base64');
    if (output !== payload) {
      console.error('Failed on', record, message);
      process.exit(1);
    }
    decompress(index + 1);
  });
}

compress(0);
```

It fails against the latest release, on record `3337`, and passes if some delay is introduced via `setTimeout` in the compression phase.

It passes on https://github.com/faye/permessage-deflate-node/commit/958de927aa925586c035332fd5a55d13e7c23a91.

**Comment by glasser** [comment]
This change looks correct.  However, two notes:

(a) I am pretty sure that the `flush: zlib.Z_FULL_SYNC` line is completely unnecessary if you're always calling flush after every write.  I would leave it out.

(b) You really really really want to pass `zlib.Z_SYNC_FLUSH` to `deflate.flush` on Node 0.12.  Node 0.8 appears to always implicitly use `Z_SYNC_FLUSH`, which is good; Node 0.10 sadly always uses `Z_FULL_FLUSH`.  `Z_FULL_FLUSH` has the same behavior of `Z_SYNC_FLUSH` of ensuring that all input to deflate ends up in the output and putting in an empty block to byte-align... but it also "forgets" all of its previous context.  It's effectively the same as using no_context_takeover (except that the inflater doesn't know that it's allowed to forget context too).

You can check this out by adding something to your script that prints the total compressed size of all messages.  On Node 0.8, the total size is 174248.  On Node 0.10 or Node 0.12, it's 1073179, over 6 times larger.  On Node 0.12, removing the `flush` option from the constructor and adding `zlib.Z_SYNC_FLUSH` as the first argument to `deflate.flush` restores the 174248 behavior again.  Following this up by configuring with `noContextTakeover` raises it back to 1073164, showing that Z_FULL_FLUSH is more or less the same as not reusing context.

Note that iojs seems to have the same behavior as Node 0.12 (but the process.version check might not work for it).  One way to detect if your `deflate` object supports the flush kind argument is to see if the method `deflate.params` exists.

**Comment by jcoglan** [comment]
With the change I just pushed, I get these results, and a clean roundtrip:

```
Now using node v0.6.21
174258
Now using node v0.8.28
174258
Now using node v0.10.38
1073179
Now using node v0.12.1
174248
Now using node iojs-v1.6.2
174248
```

With `noContextTakeover: true` set on the server session:

```
Now using node v0.6.21
1073164
Now using node v0.8.28
1073164
Now using node v0.10.38
1073164
Now using node v0.12.1
1073164
Now using node iojs-v1.6.2
1073164
```

So 0.10 actually produces _less_ output with `noContextTakeover` set. Do you know if there's any way to make `flush()` emit optimal output on 0.10?

**Comment by jcoglan** [comment]
I've run the same roundtrip script against the Ruby version and it consistently produces `174238` bytes. I wonder where the extra 10 or 20 bytes are coming from on Node?

**Comment by glasser** [comment]
It looks like the extra bytes come 5 at a time at the end of blocks and are `<Buffer 00 00 ff ff 00>`.  This matches what we've seen before: if a flushing deflate run would end at or a few bytes past the end of the output buffer, you end up with an extra block on the end.

This is unlikely to happen with noContextTakeover, because every message gets a fresh zlib stream and thus is starting its deflate calls with a fresh 16K output buffer; none of the messages we are processing will produce precisely (within 5 bytes) 16K of output.

But with `Z_FULL_FLUSH`, you're still reusing the same output buffer from call to call; you're just failing to actually create any references to past messages.  So you'll exhaust the buffer, and once in a while you'll exhaust it nearly precisely.

I don't think there's a good way to avoid the extra occasional 5 bytes, unless you literally want to look at the end of the buffer and see if it has multiple empty block encodings.

If you want to get 0.10 emitting optimal output (ie, using `Z_SYNC_FLUSH` instead of `Z_FULL_FLUSH`)... I don't think there's a clean way.  You can monkey-patch in the zlib.js change from https://github.com/joyent/node/commit/086d4ccace1d28e410f9982155d6762676d14483 or you can just plain not use zlib streams (using an alternate library or using `process.binding('zlib')` directly).  We use 0.10 but this may inspire upgrade ;)

Not super clear on where the 10 byte difference between 0.8 and 0.12 comes from.

**Comment by jcoglan** [comment]
This patch allows Node 0.10 to produce optimal output https://github.com/faye/permessage-deflate-node/commit/5da6a768b0f384923c9a0f52fa00479131d981a7

**Comment by jcoglan** [comment]
I just released version `0.1.3` with this fix.
